File: src/redis.ts
import type { DeploymentConfig } from "./config.js";
import * as pulumi from "@pulumi/pulumi";
import * as upstash from "@upstash/pulumi";

export interface RedisConnection {
  host: string;
  port: number;
  username: string;
  password: string;
  db: number;
  enableTls: boolean;
}

export interface RedisConnectionInputs {
  host: pulumi.Input<string>;
  port: pulumi.Input<number>;
  username: pulumi.Input<string>;
  password: pulumi.Input<string>;
  db: pulumi.Input<number>;
  enableTls: pulumi.Input<boolean>;
}

export function buildRedisConnection(config: DeploymentConfig): RedisConnection {
  if (config.redisMode === "docker") {
    return {
      host: "redis",
      port: 6379,
      username: config.redisUsername,
      password: config.redisPassword!,
      db: 0,
      enableTls: false,
    };
  }
  return {
    host: config.upstashHost!,
    port: config.upstashPort,
    username: config.upstashUsername,
    password: config.upstashPassword!,
    db: 0,
    enableTls: true,
  };
}

export function createUpstashConnection(
  config: DeploymentConfig,
  apiKey: pulumi.Input<string>,
): RedisConnectionInputs {
  const provider = new upstash.Provider("upstash", {
    apiKey,
    email: config.upstashEmail!,
  });
  const database = new upstash.RedisDatabase("sub2api-upstash-redis", {
    databaseName: config.upstashDatabaseName!,
    region: config.upstashRegion!,
    tls: true,
  }, { provider });

  return {
    host: database.endpoint,
    port: database.port,
    username: "default",
    password: pulumi.secret(database.password),
    db: 0,
    enableTls: true,
  };
}

File: infra/redis.go
package main

import (
	"strings"
	"github.com/pulumi/pulumi/sdk/v3/go/pulumi"
	"github.com/upstash/pulumi-upstash/sdk/go/upstash"
)

type RedisConnection struct { Host string; Port int; Username string; Password string; DB int; EnableTLS bool }
type RedisConnectionInputs struct { Host pulumi.StringInput; Port pulumi.IntInput; Username pulumi.StringInput; Password pulumi.StringInput; DB pulumi.IntInput; EnableTLS pulumi.BoolInput }
func ManagedUpstashDatabaseName(namespace, explicitName string) string { if strings.TrimSpace(explicitName) != "" { return strings.TrimSpace(explicitName) }; return namespace + "-redis" }

func siteRedisInputs(ctx *pulumi.Context, site pulumi.Resource, siteID string, spec SiteSpec, secrets SiteSecrets) (RedisConnectionInputs, error) {
	if spec.Redis.Mode == "docker" { return RedisConnectionInputs{Host: pulumi.String("redis"), Port: pulumi.Int(6379), Username: pulumi.String(""), Password: pulumi.ToSecret(pulumi.String(secrets.Redis.Password)).(pulumi.StringOutput), DB: pulumi.Int(0), EnableTLS: pulumi.Bool(false)}, nil }
	if spec.Redis.ResourceMode == "create" {
		provider, err := upstash.NewProvider(ctx, "site-"+siteID+"-upstash", &upstash.ProviderArgs{ApiKey: pulumi.ToSecret(pulumi.String(secrets.Redis.APIKey)).(pulumi.StringOutput)}, pulumi.Parent(site), pulumi.Version("0.5.0")); if err != nil { return RedisConnectionInputs{}, err }
		// Retirement must explicitly unprotect this persistent database first.
		database, err := upstash.NewRedisDatabase(ctx, "site-"+siteID+"-upstash-redis", &upstash.RedisDatabaseArgs{DatabaseName: pulumi.String(ManagedUpstashDatabaseName(spec.ResourcePrefix, "")), Region: pulumi.String(spec.Redis.Region), Tls: pulumi.BoolPtr(true)}, pulumi.Parent(site), pulumi.Provider(provider), pulumi.Version("0.5.0"), pulumi.Protect(true), pulumi.RetainOnDelete(true)); if err != nil { return RedisConnectionInputs{}, err }
		return RedisConnectionInputs{Host: database.Endpoint, Port: database.Port, Username: pulumi.String("default"), Password: pulumi.ToSecret(database.Password).(pulumi.StringOutput), DB: pulumi.Int(0), EnableTLS: pulumi.Bool(true)}, nil
	}
	endpoint, err := normalizeUpstashEndpoint(spec.Redis.Endpoint)
	if err != nil { return RedisConnectionInputs{}, err }
	return RedisConnectionInputs{Host: pulumi.String(endpoint), Port: pulumi.Int(6379), Username: pulumi.String("default"), Password: pulumi.ToSecret(pulumi.String(secrets.Redis.Password)).(pulumi.StringOutput), DB: pulumi.Int(0), EnableTLS: pulumi.Bool(true)}, nil
}
